tests/e2e/channel_init.go

```diff
@@ -93,3 +93,100 @@ func (suite *CCVTestSuite) TestProviderClientMatches() {
 	suite.Require().Equal(suite.providerClient, clientState, "stored client state does not match genesis provider client")
 }
+
+// TestInitTimeout tests the init timeout
+func (suite *CCVTestSuite) TestInitTimeout() {
+	testCases := []struct {
+		name      string
+		handshake func()
+		removed   bool
+	}{
+		{
+			"init times out before INIT", func() {}, true,
+		},
+		{
+			"init times out before TRY", func() {
+				// send ChanOpenInit
+				err := suite.path.EndpointA.ChanOpenInit()
+				suite.Require().NoError(err)
+			}, true,
+		},
+		{
+			"init times out before ACK", func() {
+				// send ChanOpenInit
+				err := suite.path.EndpointA.ChanOpenInit()
+				suite.Require().NoError(err)
+				// send ChanOpenTry
+				err = suite.path.EndpointB.ChanOpenTry()
+				suite.Require().NoError(err)
+			}, true,
+		},
+		{
+			"init times out before CONFIRM", func() {
+				// send ChanOpenInit
+				err := suite.path.EndpointA.ChanOpenInit()
+				suite.Require().NoError(err)
+				// send ChanOpenTry
+				err = suite.path.EndpointB.ChanOpenTry()
+				suite.Require().NoError(err)
+				// send ChanOpenAck
+				err = suite.path.EndpointA.ChanOpenAck()
+				suite.Require().NoError(err)
+			}, true,
+		},
+		{
+			"init completes before timeout", func() {
+				// send ChanOpenInit
+				err := suite.path.EndpointA.ChanOpenInit()
+				suite.Require().NoError(err)
+				// send ChanOpenTry
+				err = suite.path.EndpointB.ChanOpenTry()
+				suite.Require().NoError(err)
+				// send ChanOpenAck
+				err = suite.path.EndpointA.ChanOpenAck()
+				suite.Require().NoError(err)
+				// send ChanOpenConfirm
+				err = suite.path.EndpointB.ChanOpenConfirm()
+				suite.Require().NoError(err)
+			}, false,
+		},
+	}
+
+	for i, tc := range testCases {
+		providerKeeper := suite.providerApp.GetProviderKeeper()
+		initTimeout := providerKeeper.GetParams(suite.providerCtx()).InitTimeoutPeriod
+		chainID := suite.consumerChain.ChainID
+
+		// get init timeout timestamp
+		ts, found := providerKeeper.GetInitTimeoutTimestamp(suite.providerCtx(), chainID)
+		suite.Require().True(found, "cannot find init timeout timestamp; test: %s", tc.name)
+		expectedTs := suite.providerCtx().BlockTime().Add(initTimeout)
+		suite.Require().Equal(uint64(expectedTs.UnixNano()), ts, "unexpected init timeout timestamp; test: %s", tc.name)
+
+		// create connection
+		suite.coordinator.CreateConnections(suite.path)
+
+		// channel opening handshake
+		tc.handshake()
+
+		// call NextBlock
+		suite.providerChain.NextBlock()
+
+		// increment time
+		incrementTimeBy(suite, initTimeout)
+
+		// check whether the chain was removed
+		_, found = providerKeeper.GetConsumerClientId(suite.providerCtx(), chainID)
+		suite.Require().Equal(!tc.removed, found, "unexpected outcome; test: %s", tc.name)
+
+		if tc.removed {
+			// check if the chain was properly removed
+			suite.checkConsumerChainIsRemoved(chainID, false, false)
```

Comment:
Can we also check that the next step of the handshake can't complete as well

Reply:
That's not that easy due to the `expPass` param of the `SignAndDeliver` method from the ibc-go testing framework (see https://github.com/cosmos/ibc-go/blob/a0e59b8e7a2e1305b7b168962e20516ca8c98fad/testing/simapp/test_helpers.go#L336). This method is called by `SendMsgs()` with `expPass = true` (see https://github.com/cosmos/ibc-go/blob/a0e59b8e7a2e1305b7b168962e20516ca8c98fad/testing/chain.go#L319). `SendMsgs()` is called by `ChanOpenTry()`. This means that unless I change the ibc-go testing framework, calling `ChanOpenTry()` after the consumer is removed will result in the test failing with the following error: 
```
failed to execute message; message index: 0: channel open try callback failed: cannot find client for consumer chain testchain2: client not found
```

@AdityaSripal is there another way to do this?